--- postgresql-9.3beta2/src/backend/access/transam/xlog.c	2013-06-24 18:55:41.000000000 +0000
+++ postgresql-9.3beta2/src/backend/access/transam/xlog.c	2013-08-06 12:16:55.578266643 +0000
@@ -4666,7 +4666,7 @@ recoveryPausesHere(void)
 
 	while (RecoveryIsPaused())
 	{
-		pg_usleep(1000000L);	/* 1000 ms */
+		pg_usleep(10000L);	/* 10 ms */
 		HandleStartupProcInterrupts();
 	}
 }
@@ -5702,6 +5702,7 @@ StartupXLOG(void)
 
 				/* Else, try to fetch the next WAL record */
 				record = ReadRecord(xlogreader, InvalidXLogRecPtr, LOG, false);
+				SetRecoveryPause(true);
 			} while (record != NULL);
 
 			/*
